### ISSUE TITLE
no-unmodified-loop-condition: false positive plus bogus diagnostic

### ISSUE THREAD

**Issue by pabigot** [eslint/eslint#7192]
**Tell us about your environment**
- **ESLint Version:** 3.5.0
- **Node Version:** 4.5.0
- **npm Version:** 2.15.9

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```
module.exports = {
  env: {
    node: true
  },
  rules: {
    'no-unmodified-loop-condition': 'error'
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```
'use strict';

let flag;
const cb = () => {
  flag = true;
};

while (undefined === flag) {
  cb();
}
```

**What did you expect to happen?**

No diagnostics should be emitted

**What actually happened? Please include the actual, raw output from ESLint.**

```
8:8   error  'undefined' is not modified in this loop  no-unmodified-loop-condition
8:22  error  'flag' is not modified in this loop       no-unmodified-loop-condition
```

The diagnostic on an unmodified global constant property should be fixed.

I don't expect a style checker to be able to detect modification of a captured variable in a callback, so I'm just going to disable this rule that was helpfully added for me by config-eslint-xo via config-google.

**Comment by platinumazure** [comment]
The "bogus diagnostic" should go away if one of these two things happen:

1. Your code is changed so that the value of `flag` is changed in the loop. (Then the rule will report neither variable)
1. You use `void 0` instead of `undefined`, so that there are no longer two variables.

As for `flag` being flagged... I'm not sure what we can really do here. The variable is changed in the callback and that change propagates out due to how closures work, sure. But we can't know that any callback would necessarily change that variable. (What if `cb` was actually imported from another file?) So you have a couple of options there:

1. Disable the rule for this line (after all, it's technically a false positive): `// eslint-disable-line no-unmodified-loop-condition`
1. Alter the signature of your callback to return the new value and assign it in the loop so we can detect it statically within the loop body itself.

I don't know if there are any enhancements we can realistically make here. @mysticatea Any thoughts on if we can improve the rule in this situation?

**Comment by pabigot** [comment]
@platinumazure Substituting `void 0` for `undefined` makes the code rather opaque.  If somebody manages to change the value of `undefined` there's a much bigger problem, so warning that it didn't change doesn't make sense in any situation.

This rule can't be implemented completely: for it to avoid false positives while providing true positives it has to solve the halting problem.  I've disabled it globally; it was only enabled because an external style configuration did so for me, and I don't find it useful enough to warrant littering my code with directives indicating that the rule is incomplete.  I don't need a fix, just wanted to report the problem.

**Comment by mysticatea** [comment]
Actually, `no-unmodified-loop-condition` supports the situation if the arrow function is normal function declaration.

```js
'use strict';

let flag;
function cb() {
  flag = true;
}

while (undefined === flag) {  // no error.
  cb();
}
```

I think we can expand the support to function expressions and arrows.

**Comment by vitorbal** [comment]
@mysticatea given your [comment here](https://github.com/eslint/eslint/issues/7192#issuecomment-248465421), I'm marking this as a rule enhancement. Do you remember if there were any reasons why `no-unmodified-loop-condition` does not already support function expressions and arrows?

**Comment by vitorbal** [comment]
Also, @pabigot if this enhancement gets accepted, would this be something you're willing to contribute a PR for?

**Comment by pabigot** [comment]
@vitorbal sorry, no; this rule is not something I find useful.

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
